packages/in_app_purchase/ios/Classes/FIAObjectTranslator.m

```diff
@@ -83,3 +83,97 @@ - (NSDictionary *)toMap {
+                                                              encoding:NSUTF8StringEncoding]
+                                      : [NSNull null],
+    @"quantity" : @(self.quantity),
+    @"applicationUsername" : self.applicationUsername ?: [NSNull null]
+  }];
+  if (@available(iOS 8.3, *)) {
+    [map setObject:@(self.simulatesAskToBuyInSandbox) forKey:@"simulatesAskToBuyInSandbox"];
+  }
+  return map;
+}
+
+@end
+
+@implementation SKMutablePayment (Coder)
+
+- (instancetype)initWithMap:(NSDictionary *)map {
+  self = [self init];
+  if (self) {
+    self.productIdentifier = map[@"productIdentifier"];
+    NSString *utf8String = map[@"requestData"];
+    self.requestData = [utf8String dataUsingEncoding:NSUTF8StringEncoding];
+    self.quantity = [map[@"quantity"] integerValue];
+    self.applicationUsername = map[@"applicationUsername"];
+    if (@available(iOS 8.3, *)) {
+      self.simulatesAskToBuyInSandbox = [map[@"simulatesAskToBuyInSandbox"] boolValue];
+    }
+  }
+  return self;
+}
+
+@end
+
+@implementation SKPaymentTransaction (Coder)
+
+- (NSDictionary *)toMap {
+  NSMutableDictionary *map = [[NSMutableDictionary alloc] initWithDictionary:@{
+    @"error" : [self.error toMap] ?: [NSNull null],
+    @"payment" : self.payment ? [self.payment toMap] : [NSNull null],
+    @"originalTransaction" : self.originalTransaction ? [self.originalTransaction toMap]
+                                                      : [NSNull null],
+    @"transactionDate" : self.transactionDate ? @(self.transactionDate.timeIntervalSince1970)
+                                              : [NSNull null],
+    @"transactionIdentifier" : self.transactionIdentifier ?: [NSNull null],
+    @"transactionState" : @(self.transactionState)
+  }];
+  NSMutableArray *downloads = [NSMutableArray new];
+  for (SKDownload *download in self.downloads) {
+    [downloads addObject:[download toMap]];
+  }
+  [map setObject:downloads forKey:@"downloads"];
+  return map;
+}
+
+@end
+
+@implementation SKDownload (Coder)
+
+- (NSDictionary *)toMap {
+  NSMutableDictionary *map = [[NSMutableDictionary alloc] initWithDictionary:@{
+    @"contentLength" : @(self.contentLength),
+    @"contentIdentifier" : self.contentIdentifier ?: [NSNull null],
+    @"contentURL" : self.contentURL.absoluteString ?: [NSNull null],
+    @"contentVersion" : self.contentVersion ?: [NSNull null],
+    @"error" : [self.error toMap] ?: @{},
+    @"progress" : @(self.progress),
+    @"timeRemaining" : @(self.timeRemaining),
+    @"downloadTimeUnKnown" : @(self.timeRemaining == SKDownloadTimeRemainingUnknown),
+    @"transactionID" : self.transaction.transactionIdentifier ?: [NSNull null]
+  }];
+  if (@available(iOS 12.0, *)) {
+    [map setObject:@(self.state) forKey:@"state"];
+  } else {
+    [map setObject:@(self.downloadState) forKey:@"state"];
+  }
+  return map;
+}
+
+@end
+
+@implementation NSError (Coder)
+
+- (NSDictionary *)toMap {
```

Comment:
@amirh @mklim `SKDownload` and `SKTransaction` object contains an NSError as a property. Do we have a better way of sending the error object to dart or we just make it a map and let the dart side parse the map as I am doing here.